services/scan/src/precinct_scanner_state_machine.ts

```diff
@@ -0,0 +1,854 @@
+}
+
+function paperStatusErrorToEvent(
+  error: ScannerError | Error
+): ScannerStatusEvent {
+  if (error instanceof ClientDisconnectedError) {
+    return { type: 'SCANNER_DISCONNECTED' };
+  }
+  switch (error) {
+    // This error code happens when the paper gets blocked and tries to scan for
+    // a long time (or if you just hold onto the paper while scanning).
+    case ScannerError.SaneStatusInval:
+      return { type: 'SCANNER_JAM' };
+    case ScannerError.SaneStatusIoError:
+      return { type: 'SCANNER_DISCONNECTED' };
+    default:
+      throw error;
+  }
+}
+
+// Create a paper status observable, then use internal transitions to avoid
+// changing state when paper status doesn't change
+function paperStatusObserver({ client }: Context) {
+  assert(client);
+  return timer(0, 500).pipe(
+    switchMap(async () => {
+      const paperStatus = await client.getPaperStatus();
+      debugPaperStatus('Paper status: %s', paperStatus);
+      return paperStatus.isOk()
+        ? paperStatusToEvent(paperStatus.ok())
+        : paperStatusErrorToEvent(paperStatus.err());
+    })
+  );
+}
+
+const pollPaperStatus: InvokeConfig<Context, Event> = {
+  src: paperStatusObserver,
+  onError: {
+    target: 'error',
+    actions: assign({ error: (_, event) => event.data }),
+  },
+};
+
+async function scan({ client }: Context): Promise<SheetOf<string>> {
+  assert(client);
+  debug('Scanning');
+  const scanResult = await client.scan();
+  debug('Scan result: %o', scanResult);
+  const [front, back] = scanResult.unsafeUnwrap().files;
+  return [front, back];
+}
+
+async function calibrate({ client }: Context): Promise<void> {
+  assert(client);
+  debug('Calibrating');
+  const calibrateResult = await client.calibrate();
+  debug('Calibrate result: %o', calibrateResult);
+  calibrateResult.unsafeUnwrap();
+}
+
+async function interpretSheet({
+  interpreter,
+  scannedSheet,
+  interpretationMode,
+}: Context): Promise<InterpretationResultEvent> {
+  assert(interpreter);
+  assert(scannedSheet);
+
+  if (interpretationMode === 'skip') {
+    return {
+      type: 'INTERPRETATION_VALID',
+      interpretation: {
+        sheetId: 'mock-sheet-id',
+        sheet: [
+          {
+            originalFilename: '/front-original-mock',
+            normalizedFilename: '/front-normalized-mock',
+            interpretation: { type: 'BlankPage' },
+          },
+          {
+            originalFilename: '/back-original-mock',
+            normalizedFilename: '/back-normalized-mock',
+            interpretation: { type: 'BlankPage' },
+          },
+        ],
+      },
+    };
+  }
+
+  const sheetId = uuid();
+  const result = await interpreter.interpret(sheetId, scannedSheet);
+  const interpretedSheet = result.unsafeUnwrap();
+  const [front, back] = interpretedSheet;
+  const frontType = front.interpretation.type;
+  const backType = back.interpretation.type;
+  const interpretation: Interpretation = { sheetId, sheet: interpretedSheet };
+
+  if (
```

Comment:
all the interpretation stuff below this line that looks at front and back, seems like it belongs in `Interpreter`, not in the state machine.

Reply:
Agreed, that's on my todo list